libkbfs/test_stallers.go

```diff
@@ -38,6 +39,166 @@ const (
+	onStalledCh := make(chan struct{}, 1)
+	unstallCh := make(chan struct{})
+	oldBlockOps := s.config.BlockOps()
+	s.config.SetBlockOps(&stallingBlockOps{
+		stallOpName: stalledOp,
+		stallKey:    "",
+		staller: staller{
+			stalled: onStalledCh,
+			unstall: unstallCh,
+		},
+		delegate: oldBlockOps,
+	})
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.blockOpsStalls[stalledOp] = &naïveStallContext{
+		onStalled: onStalledCh,
+		unstall:   unstallCh,
+		undoStall: func() {
+			s.config.SetBlockOps(oldBlockOps)
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			delete(s.blockOpsStalls, stalledOp)
+		},
+	}
+}
+
+// StallMDOp wraps the internal MDOps so that all subsequent stalledOp
+// will be stalled. This can be undone by calling UndoStallMDOp.
+func (s *NaïveStaller) StallMDOp(stalledOp StallableMDOp) {
+	onStalledCh := make(chan struct{}, 1)
+	unstallCh := make(chan struct{})
+	oldMDOps := s.config.MDOps()
+	s.config.SetMDOps(&stallingMDOps{
+		stallOpName: stalledOp,
+		stallKey:    "",
+		staller: staller{
+			stalled: onStalledCh,
+			unstall: unstallCh,
+		},
+		delegate: oldMDOps,
+	})
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.mdOpsStalls[stalledOp] = &naïveStallContext{
+		onStalled: onStalledCh,
+		unstall:   unstallCh,
+		undoStall: func() {
+			s.config.SetMDOps(oldMDOps)
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			delete(s.mdOpsStalls, stalledOp)
+		},
+	}
+}
+
+// WaitForStallBlockOp blocks until stalledOp is stalled. StallBlockOp should
+// have been called upon stalledOp, otherwise this is a no-op.
+func (s *NaïveStaller) WaitForStallBlockOp(stalledOp StallableBlockOp) {
+	ctx := s.getNaïveStallerContextForBlockOp(stalledOp)
+	if ctx == nil {
+		return
+	}
+	<-ctx.onStalled
+}
+
+// WaitForStallMDOp blocks until stalledOp is stalled. StallMDOp should
+// have been called upon stalledOp, otherwise this is a no-op.
+func (s *NaïveStaller) WaitForStallMDOp(stalledOp StallableMDOp) {
+	ctx := s.getNaïveStallerContextForMDOp(stalledOp)
+	if ctx == nil {
+		return
+	}
+	<-ctx.onStalled
+}
+
+// UnstallOneBlockOp unstalls exactly one stalled stalledOp. StallBlockOp
+// should have been called upon stalledOp, otherwise this is a no-op.
+func (s *NaïveStaller) UnstallOneBlockOp(stalledOp StallableBlockOp) {
+	ctx := s.getNaïveStallerContextForBlockOp(stalledOp)
+	if ctx == nil {
+		return
+	}
+	ctx.unstall <- struct{}{}
+}
+
+// UnstallOneMDOp unstalls exactly one stalled stalledOp. StallMDOp
+// should have been called upon stalledOp, otherwise this is a no-op.
+func (s *NaïveStaller) UnstallOneMDOp(stalledOp StallableMDOp) {
+	ctx := s.getNaïveStallerContextForMDOp(stalledOp)
+	if ctx == nil {
+		return
+	}
+	ctx.unstall <- struct{}{}
+}
+
+// UndoStallBlockOp reverts StallBlockOp so that future stalledOp are not
+// stalled anymore.
+func (s *NaïveStaller) UndoStallBlockOp(stalledOp StallableBlockOp) {
+	ctx := s.getNaïveStallerContextForBlockOp(stalledOp)
+	if ctx == nil {
+		return
+	}
+	ctx.undoStall()
```

Comment:
Doesn't this (and the next function) need to `close(ctx.unstall)`?  It looks like `ctx.undoStall()` doesn't do that.